File: ios/Models/SessionStorage.swift
import Foundation

/// Manages saving and loading of loop sessions
final class SessionStorage {
	static let shared = SessionStorage()
	
	private let sessionsKey = "savedSessions"
	private let fileManager = FileManager.default
	
	private var sessionsURL: URL {
		let documents = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
		return documents.appendingPathComponent("sessions.json")
	}
	
	/// URL for the auto-saved working session (separate from named sessions)
	private var workingSessionURL: URL {
		let documents = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
		return documents.appendingPathComponent("working_session.json")
	}
	
	private init() {}
	
	// MARK: - Public API
	
	/// Get all saved sessions
	func loadSessions() -> [SavedSession] {
		guard fileManager.fileExists(atPath: sessionsURL.path) else {
			return []
		}
		
		do {
			let data = try Data(contentsOf: sessionsURL)
			let sessions = try JSONDecoder().decode([SavedSession].self, from: data)
			return sessions.sorted { $0.lastModifiedAt > $1.lastModifiedAt }
		} catch {
			print("❌ Failed to load sessions: \(error)")
			return []
		}
	}
	
	/// Save a new session or update existing
	func saveSession(_ session: SavedSession) {
		var sessions = loadSessions()
		
		if let index = sessions.firstIndex(where: { $0.id == session.id }) {
			// Update existing
			var updated = session
			updated.lastModifiedAt = Date()
			sessions[index] = updated
		} else {
			// Add new
			sessions.insert(session, at: 0)
		}
		
		writeSessions(sessions)
	}
	
	/// Delete a session
	func deleteSession(_ session: SavedSession) {
		var sessions = loadSessions()
		sessions.removeAll { $0.id == session.id }
		writeSessions(sessions)
	}
	
	/// Rename a session
	func renameSession(_ session: SavedSession, to newName: String) {
		var sessions = loadSessions()
		if let index = sessions.firstIndex(where: { $0.id == session.id }) {
			sessions[index].name = newName
			sessions[index].lastModifiedAt = Date()
			writeSessions(sessions)
		}
	}
	
	// MARK: - Working Session (Auto-Save)
	
	/// Save the current working session for auto-restore
	func saveWorkingSession(_ session: SavedSession) {
		do {
			let data = try JSONEncoder().encode(session)
			try data.write(to: workingSessionURL)
			print("✓ Auto-saved working session")
		} catch {
			print("❌ Failed to auto-save working session: \(error)")
		}
	}
	
	/// Load the auto-saved working session (if any)
	func loadWorkingSession() -> SavedSession? {
		guard fileManager.fileExists(atPath: workingSessionURL.path) else {
			return nil
		}
		
		do {
			let data = try Data(contentsOf: workingSessionURL)
			let session = try JSONDecoder().decode(SavedSession.self, from: data)
			return session
		} catch {
			print("❌ Failed to load working session: \(error)")
			return nil
		}
	}
	
	/// Clear the auto-saved working session
	func clearWorkingSession() {
		guard fileManager.fileExists(atPath: workingSessionURL.path) else {
			return
		}
		
		do {
			try fileManager.removeItem(at: workingSessionURL)
			print("✓ Cleared working session")
		} catch {
			print("❌ Failed to clear working session: \(error)")
		}
	}
	
	// MARK: - Private
	
	private func writeSessions(_ sessions: [SavedSession]) {
		do {
			let data = try JSONEncoder().encode(sessions)
			try data.write(to: sessionsURL)
			print("✓ Saved \(sessions.count) sessions")
		} catch {
			print("❌ Failed to save sessions: \(error)")
		}
	}
}

File: android-app/core/storage/src/main/java/com/loopa/core/storage/SessionStorage.kt
package com.loopa.core.storage

import android.content.Context
import com.loopa.core.model.SavedSession
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/**
 * Manages saving and loading of loop sessions.
 * Ported from iOS SessionStorage.swift.
 *
 * Date encoding note:
 * - Named sessions (sessions.json) use default epoch milliseconds
 * - Working session (working_session.json) uses default epoch milliseconds
 * - .loopa export uses ISO-8601 (handled by SessionExporter)
 */
class SessionStorage(private val context: Context) {

    private val json = Json {
        prettyPrint = false
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    private val sessionsFile: File
        get() = File(context.filesDir, "sessions.json")

    private val workingSessionFile: File
        get() = File(context.filesDir, "working_session.json")

    // MARK: - Named Sessions

    /** Get all saved sessions, sorted by lastModifiedAt descending */
    fun loadSessions(): List<SavedSession> {
        if (!sessionsFile.exists()) return emptyList()
        return try {
            val data = sessionsFile.readText()
            val sessions = json.decodeFromString<List<SavedSession>>(data)
            sessions.sortedByDescending { it.lastModifiedAt }
        } catch (e: Exception) {
            emptyList()
        }
    }

    /** Save a new session or update existing */
    fun saveSession(session: SavedSession) {
        val sessions = loadSessions().toMutableList()
        val index = sessions.indexOfFirst { it.id == session.id }
        if (index >= 0) {
            sessions[index] = session.copy(lastModifiedAt = System.currentTimeMillis())
        } else {
            sessions.add(0, session)
        }
        writeSessions(sessions)
    }

    /** Delete a session */
    fun deleteSession(session: SavedSession) {
        val sessions = loadSessions().toMutableList()
        sessions.removeAll { it.id == session.id }
        writeSessions(sessions)
    }

    /** Rename a session */
    fun renameSession(session: SavedSession, newName: String) {
        val sessions = loadSessions().toMutableList()
        val index = sessions.indexOfFirst { it.id == session.id }
        if (index >= 0) {
            sessions[index] = sessions[index].copy(
                name = newName,
                lastModifiedAt = System.currentTimeMillis()
            )
            writeSessions(sessions)
        }
    }

    // MARK: - Working Session (Auto-Save)

    /** Save the current working session for auto-restore */
    fun saveWorkingSession(session: SavedSession) {
        try {
            val data = json.encodeToString(session)
            workingSessionFile.writeText(data)
        } catch (e: Exception) {
            // Log error
        }
    }

    /** Load the auto-saved working session (if any) */
    fun loadWorkingSession(): SavedSession? {
        if (!workingSessionFile.exists()) return null
        return try {
            val data = workingSessionFile.readText()
            json.decodeFromString<SavedSession>(data)
        } catch (e: Exception) {
            null
        }
    }

    /** Clear the auto-saved working session */
    fun clearWorkingSession() {
        if (workingSessionFile.exists()) {
            workingSessionFile.delete()
        }
    }

    // MARK: - Private

    private fun writeSessions(sessions: List<SavedSession>) {
        try {
            val data = json.encodeToString(sessions)
            sessionsFile.writeText(data)
        } catch (e: Exception) {
            // Log error
        }
    }
}
